Fix long multibyte characters paste issue via ssh

akifumi9 opened the pull request:
When I paste long multibyte characters(over 80 byte) to ksh via SSH, the characters are not displayed correctly.
For example, the following input demonstrates the issue.  ja_JP.UTF-8 encoding is used.
input  : echo "長い文字列を入れるとkshで文字列が乱れる場合があるようです"
output: です"echo "長い文字列を入れるとkshで文字列が乱れる場合がある

This issue appears to be caused by the ed_read() function splitting a multibyte character sequence when reading into an 80-byte buffer.
This leads to incorrect character interpretation and display.

Therefore, we edited the code to handle the case where the buffer size is full in the middle of a multi-byte character.

akifumi9 committed:
Fix long multibyte characters paste issue via ssh

Fix edit.c
 Fixed the handling of cases where the parsing of multi-byte characters
 does not fit into the buffer.

McDutchie commented:
Many thanks for not only finding the bug, but fixing it! The PR looks good to me.

I'll just add some housekeeping stuff, and get rid of the obsolete `#ifdef EILSEQ` while we're at it (it's defined in the C99 standard; every system from the last quarter century will have it).

McDutchie commented on src/cmd/ksh93/edit/edit.c:
```diff
@@ -828,7 +834,7 @@ static int putstack(Edit_t *ep,char string[], int nbyte, int type)
 {
 	int n = 0, c;
-	char readin[LOOKAHEAD+1];
+	char readin[LOOKAHEAD+6];
```
On second thought: while this is good for UTF-8, which has a maximum character length of 6 bytes, it will break in locale that may have longer characters. In fact, libast's DEBUG locale has a maximum character length of 7 bytes.

To be future-proof, I think it's better to allocate this dynamically based on the current maximum length, accessed via the mbmax() macro from ast.h.

```c
        static char *readin;
        static int old_mbmax;
        if(old_mbmax < mbmax())
                readin = sh_realloc(readin, LOOKAHEAD + (old_mbmax = mbmax()));
```

akifumi9 replied:
Thank you for the review.
I think it is better to allocate memory dynamically as you mentioned.
I did a quick test on my environment and I think this code is fine.

McDutchie replied:
On third thought, it would be much simpler to just use `fmtbuf` from libast.

```c
char *readin = fmtbuf(LOOKAHEAD + mbmax());
```

McDutchie committed:
edit.c tweaks

- get rid of obsolete #ifdef EILSEQ
- ed_getchar(): dynamically (re)allocate readin on the heap to add
  mbmax() bytes even if mbmax() grows

McDutchie committed:
housekeeping

McDutchie committed:
ed_getcharf(): use fmtbuf instead

McDutchie committed:
bump release date

McDutchie committed:
putstack(): block-local scope for prevp
